File: LeetcodePractices/Simple/Concurrency/Print in order/Python/printinorder.py
class Foo(object):
    def __init__(self):
        self.lock1 = threading.Lock()
        self.lock2 = threading.Lock()
        self.lock3 = threading.Lock()
        pass


    def first(self, printFirst):
        """
        :type printFirst: method
        :rtype: void
        """
        
        # printFirst() outputs "first". Do not change or remove this line.
        self.lock1.acquire()
        printFirst()
        self.lock1.release()


    def second(self, printSecond):
        """
        :type printSecond: method
        :rtype: void
        """
        
        # printSecond() outputs "second". Do not change or remove this line.
        self.lock1.acquire()
        self.lock2.acquire()
        printSecond()
        self.lock1.release()
        self.lock2.release()
            
            
    def third(self, printThird):
        """
        :type printThird: method
        :rtype: void
        """
        
        # printThird() outputs "third". Do not change or remove this line.
        self.lock1.acquire()
        self.lock2.acquire()
        self.lock3.acquire()
        printThird()
        self.lock1.release()
        self.lock2.release()
        self.lock3.release()

File: LeetcodePractices/Simple/Concurrency/Print in order/C/printinorder.c
#include "pthread.h"

typedef struct {
// User defined data may be declared here.
pthread_mutex_t first_lock;
pthread_mutex_t next_lock;

} Foo;

Foo* fooCreate() {
Foo* obj = (Foo*) malloc(sizeof(Foo));
pthread_mutex_init(&obj->first_lock,NULL);
pthread_mutex_init(&obj->next_lock,NULL);
pthread_mutex_lock(&obj->first_lock);
pthread_mutex_lock(&obj->next_lock);
// Initialize user defined data here.
return obj;
}

void first(Foo* obj) {

// printFirst() outputs "first". Do not change or remove this line.
printFirst();
pthread_mutex_unlock(&obj->first_lock);
}

void second(Foo* obj) {
pthread_mutex_lock(&obj->first_lock);
// printSecond() outputs "second". Do not change or remove this line.
printSecond();
pthread_mutex_unlock(&obj->next_lock);
}

void third(Foo* obj) {
pthread_mutex_lock(&obj->next_lock);
// printThird() outputs "third". Do not change or remove this line.
printThird();
}

void fooFree(Foo* obj) {
// User defined data may be cleaned up here.
free(obj);
}